homeassistant/helpers/template.py

```diff
@@ -1042,5 +1043,10 @@ def is_safe_attribute(self, obj, attr, value):
         """Test if attribute is safe."""
         return isinstance(obj, Namespace) or super().is_safe_attribute(obj, attr, value)
 
+    @lru_cache(maxsize=_TEMPLATE_CACHE_SIZE_PER_ENV)
```

Comment:
Instead of lru cache, wouldn't it make more sense to use a [`weakref.WeakValueDictionary`](https://docs.python.org/3/library/weakref.html#weakref.WeakValueDictionary)? When we do a reload, the templates are still in memory and so we should be able to reuse them? That way we always have all templates that are still in use.

Reply:
👍

Will give that a try and compare the hit and miss rates to make sure there isn't a significant difference.

Reply:
A few more cache misses, but I'm not sure these should be run though the template engine anyways
```
2020-08-03 18:04:16 WARNING (MainThread) [homeassistant.helpers.template] Cache miss: New devices discovered
2020-08-03 18:04:16 WARNING (MainThread) [homeassistant.helpers.template] Cache miss: We have discovered new devices on your network. [Check it out](/config/integrations)
2020-08-03 18:04:16 WARNING (MainThread) [homeassistant.helpers.template] Cache hit: New devices discovered
2020-08-03 18:04:16 WARNING (MainThread) [homeassistant.helpers.template] Cache hit: We have discovered new devices on your network. [Check it out](/config/integrations)
2020-08-03 18:04:16 WARNING (MainThread) [homeassistant.helpers.template] Cache miss: New devices discovered
2020-08-03 18:04:16 WARNING (MainThread) [homeassistant.helpers.template] Cache miss: We have discovered new devices on your network. [Check it out](/config/integrations)
2020-08-03 18:04:16 WARNING (MainThread) [homeassistant.helpers.template] Cache hit: New devices discovered
2020-08-03 18:04:16 WARNING (MainThread) [homeassistant.helpers.template] Cache hit: We have discovered new devices on your network. [Check it out](/config/integrations)
2020-08-03 18:04:16 WARNING (MainThread) [homeassistant.helpers.template] Cache miss: New devices discovered
2020-08-03 18:04:16 WARNING (MainThread) [homeassistant.helpers.template] Cache miss: We have discovered new devices on your network. [Check it out](/config/integrations)
2020-08-03 18:04:16 WARNING (MainThread) [homeassistant.helpers.template] Cache miss: New devices discovered
2020-08-03 18:04:16 WARNING (MainThread) [homeassistant.helpers.template] Cache miss: We have discovered new devices on your network. [Check it out](/config/integrations)
2020-08-03 18:04:23 WARNING (MainThread) [homeassistant.helpers.template] Cache miss: New devices discovered
2020-08-03 18:04:23 WARNING (MainThread) [homeassistant.helpers.template] Cache miss: We have discovered new devices on your network. [Check it out](/config/integrations)
```

Reply:
Still getting high cache hit ratios on automations and scripts 👍

Reply:
Converted as suggested